### PR TITLE
refactor(Progress): replaced class components by functional components

### PR DESCRIPTION
**Pull request by atroynikov**
* splited to multiple simple components
* added React.forwardRef

### CODE REVIEW

File: src/components/Progress/ProgressInnerText.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+
+import {block} from '../utils/cn';
+
+const b = block('progress');
```
Comment (Lunory) [inline src/components/Progress/ProgressInnerText.tsx:5]: Should move block init to constants like https://github.com/gravity-ui/uikit/blob/e9c092db83b3bdc76e5ab1586632a51b7e535348/src/components/Sheet/constants.ts

File: src/components/Progress/ProgressInnerText.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+export interface ProgressInnerTextProps {
+    text?: string;
+    offset?: number;
+}
+
+export function ProgressInnerText(props: ProgressInnerTextProps) {
+    const {text, offset} = props;
+
+    if (!text || typeof offset === 'undefined') {
```
Comment (Lunory) [inline src/components/Progress/ProgressInnerText.tsx:15]: `|| -> &&`?

File: src/components/Progress/ProgressStackItem.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        loading,
+    };
+
+    if (typeof color === 'undefined') {
+        modifiers.theme = theme || 'default';
+    }
+
+    if (Number.isFinite(value)) {
+        return (
+            <div
```
Comment (Lunory) [inline src/components/Progress/ProgressStackItem.tsx:26]: Missed `key={index}`

File: src/components/Progress/ProgressStackItem.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    const {value, color, className, theme, title, content, loading} = item;
+    const modifiers: CnMods = {
+        loading,
+    };
+
+    if (typeof color === 'undefined') {
+        modifiers.theme = theme || 'default';
+    }
+
+    if (Number.isFinite(value)) {
```
Comment (Lunory) [inline src/components/Progress/ProgressStackItem.tsx:24]: Let's add check for `!isNaN(value)` like it was in class component

Hmm, maybe move this check to utils for reusing?

File: src/components/Progress/ProgressWithValue.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import type {ProgressWithValueProps} from './types';
+import {getOffset, getTheme} from './utils';
+
+const b = block('progress');
+
+export function ProgressWithValue(props: ProgressWithValueProps) {
+    const {value, loading, text} = props;
+    const offset = getOffset(value);
+
+    if (!Number.isFinite(value)) {
```
Comment (Lunory) [inline src/components/Progress/ProgressWithValue.tsx:15]: same `!isNaN(value)`

File: src/components/Progress/ProgressStackItem.tsx (hunk 4 above)
Comment (atroynikov) [reply src/components/Progress/ProgressStackItem.tsx:24]: The `Number.isFinite` static method determines whether the passed value is a finite number — that is, it checks that a given value is a number, and the number is neither positive Infinity, negative Infinity, nor NaN.

File: src/components/Progress/ProgressStackItem.tsx (hunk 3 above)
Comment (atroynikov) [reply src/components/Progress/ProgressStackItem.tsx:24]: https://github.com/gravity-ui/uikit/pull/1214/files#diff-6f705be07741a801b28ec7d67bf05d20fc1243aec6794d7d0f44b5d9425d46c8R20

**Comment by Lunory** [comment]
Thanks, now that's OK

Finally, can you change Progress README? Add info about`ref` and remove `loading` prop default value

**Comment by atroynikov** [comment]
> Thanks, now that's OK
> 
> Finally, can you change Progress README? Add info about`ref` and remove `loading` prop default value

I returned initialisation for all default values from docs. There is no info about ref property in other components.

**Pull request merged**